### PR TITLE
Rename `Chain` -> `Stack`

### LINKED ISSUE #233: Consider renaming `Chain` -> `Stack`
"chain" is mixing metaphors.

### PR DESCRIPTION
**Pull request by carllerche**
Try to reduce metaphores.

Fixes #233

### CODE REVIEW

File: tower/src/layer.rs [hunk 1]
```diff
@@ -14,12 +14,12 @@ pub trait LayerExt<S, Request>: Layer<S, Request> {
     /// `middleware` to services being wrapped.
     ///
     /// This defines a middleware stack.
-    fn chain<T>(self, middleware: T) -> util::Chain<Self, T>
+    fn chain<T>(self, middleware: T) -> util::Stack<Self, T>
```
Comment (hawkw) [inline tower/src/layer.rs]: It's possible we've already discussed this, but I feel like the type and method ought to share the same name?
```suggestion
    fn stack<T>(self, middleware: T) -> util::Stack<Self, T>
```

Comment (LucioFranco) [reply tower/src/layer.rs]: 👍 on this

Comment (carllerche) [reply tower/src/layer.rs]: Ah yes... but I was thinking about getting rid of that fn. I don't think it is needed w/ `ServiceBuilder`.

Comment (carllerche) [reply tower/src/layer.rs]: Thanks for catching this though.

Comment (LucioFranco) [reply tower/src/layer.rs]: I still think it could be useful 🤷‍♂️ I don't see why there is a reason to remove it over keeping it.

**Comment by carllerche** [comment]
I'm going to just remove `LayerExt`.

File: tower/src/layer.rs (hunk 1 above)
Comment (carllerche) [reply tower/src/layer.rs]: Can always bring it back later. I rather stay on the conservative side initially.

**Comment by LucioFranco** [comment]
Not sure if there is a benefit to removing it? I personally can see people maybe using it 🤷‍♂️

File: tower/src/layer.rs (hunk 1 above)
Comment (carllerche) [reply tower/src/layer.rs]: There is only 1 fn right now... `Stack::new` can be called directly, and `ServiceBuilder` handles the use case too.

**Comment by LucioFranco** [comment]
but if removing it means that we might add it later then im fine with that.

**Pull request merged**